Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py

```diff
@@ -160,42 +180,89 @@ def _oproxy_authorize(self) -> Tuple[str, int, str]:
         }
 
         if self.scope:
             data['scope'] = self.scope
         if self.resource:
             data['resource'] = self.resource
 
-        body: dict = {}
+        response_json: dict = {}
         try:
-            response = requests.post(url, data, verify=self.verify)
+            response = requests.post(self.token_retrieval_url, data, verify=self.verify)
             if response.status_code not in {200, 201}:
                 return_error(f'Error in Microsoft authorization. Status: {response.status_code},'
                              f' body: {self.error_parser(response)}')
-            body = response.json()
+            response_json = response.json()
         except Exception as e:
             return_error(f'Error in Microsoft authorization: {str(e)}')
 
-        access_token = body.get('access_token', '')
-        expires_in = int(body.get('expires_in', 3595))
+        access_token = response_json.get('access_token', '')
+        expires_in = int(response_json.get('expires_in', 3595))
+
+        return access_token, expires_in, ''
+
+    def _get_self_deployed_token_auth_code(self, refresh_token: str = '') -> Tuple[str, int, str]:
+        """
+        Gets a token by authorizing a self deployed Azure application.
+
+        Returns:
+            tuple: An access token, its expiry and refresh token.
+        """
+        data = {
+            'client_id': self.client_id,
+            'client_secret': self.client_secret,
+            'resource': self.resource,
+            'redirect_uri': self.redirect_uri
+        }
+        refresh_token = refresh_token or self._get_refresh_token_from_auth_code_param()
+        if refresh_token:
+            data['grant_type'] = REFRESH_TOKEN
+            data['refresh_token'] = refresh_token
+        else:
+            data['grant_type'] = AUTHORIZATION_CODE
+            data['code'] = self.auth_code
+
+        response_json: dict = {}
+        try:
+            response = requests.post(self.token_retrieval_url, data, verify=self.verify)
+            if response.status_code not in {200, 201}:
+                return_error(f'Error in Microsoft authorization. Status: {response.status_code},'
+                             f' body: {self.error_parser(response)}')
+            response_json = response.json()
+        except Exception as e:
+            return_error(f'Error in Microsoft authorization: {str(e)}')
+
+        access_token = response_json.get('access_token', '')
+        # refresh_token = response_json.get('refresh_token', '')
+        expires_in = int(response_json.get('expires_in', 3595))
+
+        return access_token, expires_in, refresh_token
 
-        return access_token, expires_in
+    def _get_refresh_token_from_auth_code_param(self) -> str:
+        refresh_prefix = "refresh_token:"
+        if self.auth_code.startswith(refresh_prefix):  # for testing we allow setting the refresh token directly
+            demisto.debug("Using refresh token set as auth_code")
```

Comment:
not sure i understand these two lines

Reply:
It lets you add a refresh token under `auth_code` parameter in the format `refresh_token:<token>`. That way you can test the self deployed method using refresh token (you can't use an auth code - it's a one-time code, you can use it only once, and you can save refresh tokens because you don't have an integration context).